### PR TITLE
Infer event types

### PR DESCRIPTION
**Pull request by Velenir**
With Typescript 4.1 it's possible to infer types from  string template

So given abi string `'event Transfer(address indexed from, address indexed to, uint amount)'` we can automatically infer 
```ts
interface TransferParams {
  from: string
  to: string
  amount: BigNumber
}
```

This PR

-  updates TS to latest
- fixes broken types (due to update)
- adds type inference from events (for simple types, not structs,nor arrays)

### CODE REVIEW

File: package.json
Comment (anxolin) [inline package.json:89]: My only concern dima, is that we will use a different version as Uniswap for getting this fancy feature. I would be worried if this breaks anything in the future or adds a the need to "hack" pieces of their code to migrate to this version of TS.

**Review by anxolin — APPROVED** [review]
I approve, because obviously you went through some trouble to make this fancy feature of TS work, and honestly is cool.

However, I have some concerns of applying this into this project, since you are changing TS version and even modifying some of the original files.

Please, only merge if you think this won't explode to us and bring us trouble. It's not that bad to define the TS (or autogenerate) for this project so we keep as close to Uni as possible.
I leave it up to you

**Comment by Velenir** [comment]
> However, I have some concerns of applying this into this project, since you are changing TS version and even modifying some of the original files.

Valid concerns. And I don't feel like fighting with upstream conflicts.
I'll mark as experimental and put this PR on hold

**Review by W3stside — APPROVED** [review]
Same comment as @anxolin - great work, looks cool, but S C A R Y

**Pull request closed without merging**
